proposals/span-safety.md

```diff
@@ -0,0 +1,188 @@
+**A person who does not intend to trade in stack-referring spans should be able to stop reading right here.**
+ 
+NOTE: nothing will prevent a possibility to use Spans that wrap pointers to stack-allocated arrays in _unsafe_ code as long as it is possible to wrap a pointer in a span. As per usual rules of _unsafe_, it would be responsibility of the user to not cause GC holes.
+
+
+## Stack-referring spans in safe code.
+
+There are two main scenarios where stack-referring spans could come into being.  
+- Stackalloc spans.
+  To enable safe wrapping of stack allocated memory in a span, we will need a special syntax that would not allow intermediate pointer to escape.
+  A preliminary solution is to use target-typing.
+
+```cs
+  //This usage of stackalloc does not require unsafe 
+  Span<int> sp = stackalloc int[100]; 
+```
+  
+- Spans wrapping individual locals or parameters. [Suggest disallowing]
+ 
+```cs
+  int x = 42;
+
+  //This is currently not possible
+  Span<int> sp = new Span<int>(ref x);
+```
+
+  This is currently not possible in Span API. I suggest to never allow. 
+
+  While in theory possible, one-element spans are not as practically appealing as stackalloc ones.
+
+  One problem is that in addition to being tied to the stack frame, local variables have scopes, so we would need extra rules to prevent mixing locals of different life-times when passing by reference or assigning to other locals.
+ 
+  The main problem is the "pay-for-play" rule. If we have to assume that any span value returned from a member that takes `ref/out/in` parameters, including `this` in struct members might be stack-referring. People that do not use stack-referring spans will be affected by this.
+
+```cs
+  int x = 42; 
+
+  //Do I really have a stack-referring span here?
+  Span<int> sp = SomeMethod(ref x); 
+
+```
+
+## Safety rules for stack-referring spans.
+Note: all these rules equally apply to ref-like types in general. I just use span for brevity.
+
+- stack-referring span values cannot be returned from a method.
+
+- local variable initialized using a stack-referring expression is considered a stack-referring local.
+
+- stack-referring local is conservatively considered as always returning a stack-referring value.
+
+- stack-referring values cannot be assigned to any variable with exception of stack-referring locals.
+
+- "What comes in is what comes out" rule for span-returning expressions:
+  If any operand of an expression is a stack-referring value, the result is a stack-referring value.
+  This applies equally to ternary expressions, indexers, method calls, and so on. 
+  Receiver of an invocation/property/indexer is considered an operand here.
+ 
+  In particular: _Slice of a stack-referring value is a stack-referring value._  
+
+- "No Mixing with refs to ordinary" rule: 
+  Stack-referring spans can be passed as an argument as long as no other argument is an ordinary span variable (not a stack-referring local), and:
```

Comment:
What's the rationale for this? It's not clear to me (not being involved in this, why this rule exists)